impl/src/main/java/com/okta/sdk/impl/http/authc/OAuth2RequestAuthenticator.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2020 Okta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.okta.sdk.impl.http.authc;
+
+import com.okta.commons.http.Request;
+import com.okta.commons.http.authc.RequestAuthenticationException;
+import com.okta.commons.http.authc.RequestAuthenticator;
+import com.okta.sdk.authc.credentials.ClientCredentials;
+import com.okta.sdk.impl.oauth2.OAuth2AccessToken;
+import com.okta.sdk.impl.oauth2.OAuth2ClientCredentials;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.security.InvalidKeyException;
+
+public class OAuth2RequestAuthenticator implements RequestAuthenticator {
+    private static final Logger log = LoggerFactory.getLogger(OAuth2RequestAuthenticator.class);
+
+    public static final String AUTHENTICATION_SCHEME = "OAUTH2";
+
+    private final ClientCredentials<OAuth2AccessToken> clientCredentials;
+
+    public OAuth2RequestAuthenticator(ClientCredentials<OAuth2AccessToken> clientCredentials) {
+        this.clientCredentials = clientCredentials;
+    }
+
+    @Override
+    public void authenticate(Request request) throws RequestAuthenticationException {
```

Comment:
There are a few problems with this method.

1.) it doesn't look thread-safe
2.) it checks if the token has expired, but doesn't handle cases when 401/403 is returned from Okta if the token has been invalidated on the server, or a time sync issue that would cause the client to use an expired token that the server will reject. (this starts getting into edge cases, so check with eng team to see what types of errors we could see, and add UTs for them)

Reply:
1. Noted, I'll address this thread safety issue.
2. Yes, it gets tricky as we think about more possible edge cases. I'll document the possible scenarios and talk to product/engg and see if we can work on them as a new JIRA task later. What do you think?

Reply:
Re #2 
It depends on what they say.  This is the type of issue that if it were to pop up it would be very difficult to figure out the problem.

If we need to push it off, we can remove the references to this feature in the README until those were in.  That would allow us to get the code and start testing it, but it also wouldn't look like we felt it was production ready.

Reply:
#1 is fixed
#2 working on it.

Reply:
@bdemers & I have discussed #2 in depth and decided to document the caveats in README.